File: 2023/day09/day09.py
import re

def allSame(arr: list) -> bool:
    if len(arr) == 0:
        return True
    
    return all([num == arr[0] for num in arr[1:]])


def main():
    with open("input.txt") as f:
        lines = f.readlines()

    sum = 0
    for line in lines:
        nums = [int(x) for x in re.findall(r"-?\d+", line)]

        diffs = nums
        firsts = [nums[0]]
        while not allSame(diffs):
            prev = diffs
            diffs = []
            for i in range(len(prev) - 1):
                diffs.append(prev[i + 1] - prev[i])

            print(diffs)
            firsts.append(diffs[0])

        curr = 0
        for num in reversed(firsts):
            curr = num - curr

        print(curr)
        sum += curr

    print(sum)


if __name__ == "__main__":
    main()

File: 2023/day09/day09.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func extractNums(line string) []int {
	r, _ := regexp.Compile(`-?\d+`)
	numsStrings := r.FindAllString(line, -1)
	nums := make([]int, 0)

	for _, numString := range numsStrings {
		val, _ := strconv.Atoi(numString)
		nums = append(nums, val)
	}

	return nums
}

func allSame(nums []int) bool {
	if len(nums) == 0 {
		return true
	}

	first := nums[0]
	for _, num := range nums[1:] {
		if num != first {
			return false
		}
	}

	return true
}

func main() {
	body, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Printf("failed to read file: %v\n", err.Error())
		return
	}

	sum := 0
	lines := strings.Split(strings.Trim(string(body), "\n"), "\n")
	for _, line := range lines {
		nums := extractNums(line)
		lastNums := make([]int, 0)
		diffs := nums
		fmt.Println(nums)
		lastNums = append(lastNums, nums[len(nums)-1])
		for !allSame(diffs) {
			prev := diffs
			diffs = make([]int, len(diffs)-1)
			for i := range prev[1:] {
				diffs[i] = prev[i+1] - prev[i]
			}

			fmt.Println(diffs)
			lastNums = append(lastNums, diffs[len(diffs)-1])
		}

		total := 0
		for _, num := range lastNums {
			total += num
		}

		fmt.Println(total)
		sum += total
	}

	fmt.Println(sum)
}
